### PR TITLE
[dtensor][experimental] local_map

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123676

**Summary**
This PR is attempt to land an experimental feature designed in #103686 . `local_map` is designed to allow users to apply to `DTensor` objects a function that was written to apply to `torch.Tensor`.

As a function, `local_map` takes in 2 required arguments (`func` and `out_placements`) and 3 optional arguments (`device_mesh`, `in_placements`, `redistribute_inputs`). `func` is the function to be applied to each local shard of input `DTensor`. `out_placements` is the sharding specification of output `DTensor`.

`local_map` returns a new function that does the following:

1. Infer `device_mesh` and `in_placements` from `DTensor` input if they're not provided. If `device_mesh` is provided, it must be identical to the device mesh of every `DTensor` input. If `in_placements` is provided, it serves as the required sharding specification of corresponding `DTensor` input before feeding its local shard into `func`. In case it is different from `DTensor`'s sharding specification, if `redistribute_inputs=False` an exception will be raised, otherwise perform a resharding to the required sharding.
2. Call `func` with the arguments passed in along with `device_mesh` except `DTensor`s. For `DTensor`, pass in its local shard. This `func` may include collectives.
3. For each output of `func` that has validate (i.e. not `None) sharding specification in `out_placements`, construct a new `DTensor` using the output and the specification. Use this `DTensor` as the output.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/_tensor/experimental/dmap.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+PlacementType = Sequence[Placement]
+InputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+OutputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def dmap(
+    func: Callable,
+    out_placements: OutputPlacements,
+    device_mesh: Optional[DeviceMesh] = None,
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/dmap.py:18]: if both `out_placements` and `in_placements` can be None, let's put device_mesh after in_placements

File: torch/distributed/_tensor/experimental/dmap.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                out_placements[idx]
+                if isinstance(out_placements, tuple)
+                else out_placements
+            )
+            if isinstance(out, torch.Tensor):
+                assert not isinstance(
+                    out, DTensor
+                ), f"torch.Tensor output expected but received {type(out)}: {out}"
+
+                flat_dist_out.append(DTensor.from_local(out, device_mesh, spec))
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/dmap.py:145]: from local should call with `run_check=False`, as user mark the out placement sharding we should assume they are correct

File: torch/distributed/_tensor/experimental/dmap.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    Args:
+        func (Callable): the function to be applied on each local shard of :class:`DTensor`s.
+        out_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]]):
+            the desired placements of the output :class:`DTensor`s.
+        device_mesh (:class:`DeviceMesh`, optional):
+            the device mesh that all the :class:`DTensor`s are placed on. If not specified,
+            this will be inferred from the input :class:`DTensor`s' device mesh. `dmap` requires
+            every :class:`DTensor`s to be placed on the same device mesh.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/dmap.py:34]: I think we should explicitly mention that the in_placements are out placements are flattened list if the function's actual inputs/outputs are complicated data structures like nested tuples or dict?

File: torch/distributed/_tensor/experimental/dmap.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    NOTE: This API is currently experimental and subject to change
+    """
+
+    def wrapped(*args, **kwargs):
+        # process input args
+        flat_args, args_spec = tree_flatten(args)
+
+        # we assume every DTensor object is placed on the same device mesh
+        flat_local_args = []
+        nonlocal device_mesh  # access var device_mesh from the outer scope
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/dmap.py:91]: what does this mean? I thought device_mesh from the dmap API should be visible here even without nonlocal?

**Review by wanchaol — COMMENTED** [review]
first pass, looks pretty good already, please see inlined comments.

File: torch/distributed/_tensor/experimental/dmap.py (hunk 4 above)
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/dmap.py:91]: The reason we need `nonlocal` here is that `device_mesh` may be updated if it's passed in as `None`. Without `nonlocal`, the assignment `device_mesh = arg.device_mesh` will be treated as creating a new local variable `device_mesh` and initializing it to `arg.device_mesh`. If without any `device_mesh = ...` assignment, we would not need the `nonlocal` keyword.

**Review by tianyu-l — COMMENTED** [review]
Very interesting work! Left some inline comments.

File: torch/distributed/_tensor/experimental/dmap.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/dmap.py:145]: Good question! Generally in dtensor we don't check the type of `spec` because in Python without individually checking every element we can only assert if `spec` itself is a `Sequence`. So usually we just check if `spec` is `None` or not.

**Comment by XilunWu** [comment]
`local_map` should also be able to take in `kwargs` placements as in https://github.com/pytorch/pytorch/pull/124114

File: test/distributed/_tensor/experimental/test_local_map.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def mm_forward(device_mesh, W, X):
+    return torch.mm(W, X)
+
+
+def mm_allreduce_forward(device_mesh, W, X):
+    partial_sum_tensor = torch.mm(W, X)
+    reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+    return reduced_tensor
```
Comment (wanchaol) [inline test/distributed/_tensor/experimental/test_local_map.py:30]: nit: maybe explicitly wait for this collective call, otherwise this would leak AsyncColelctiveTensor out

File: test/distributed/_tensor/experimental/test_local_map.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+)
+from torch.distributed._tensor.debug import CommDebugMode
+from torch.distributed._tensor.experimental import local_map
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+)
+
+DEVICE_TYPE = "cpu"
```
Comment (wanchaol) [inline test/distributed/_tensor/experimental/test_local_map.py:20]: is there a particular reason to only test this in CPU? Please change this to align with other tests so that we can test in multi gpus

File: torch/distributed/_tensor/experimental/local_map.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        >>>     reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+        >>>     return reduced_tensor
+        >>>
+        >>> W = torch.randn(12, 8, requires_grad=False)
+        >>> X = torch.randn(8, 16, requires_grad=False)
+        >>> Y = torch.mm(W, X)
+        >>> row_wise = [Shard(0)]  # row-wise sharding placements on 1-d mesh
+        >>> col_wise = [Shard(1)]  # col-wise sharding placements on 1-d mesh
+        >>>
+        >>> # d_mm_allreduce_forward is the function wrapped with DTensor/Tensor convertion
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/local_map.py:83]: nit: we should call this `local_mm_allreduce_forward`

File: torch/distributed/_tensor/experimental/local_map.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        # process input args
+        flat_args, args_spec = tree_flatten(args)
+
+        # we assume every DTensor object is placed on the same device mesh
+        flat_local_args = []
+        nonlocal device_mesh  # access var device_mesh from the outer scope
+        for idx, arg in enumerate(flat_args):
+            if isinstance(arg, DTensor):
+                if device_mesh is None:  # infer device mesh from the DTensor arg
+                    device_mesh = arg.device_mesh
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/local_map.py:108]: Can you add a TODO here? IMO we should probably investigate how this would interact with uneven sharding

File: test/distributed/_tensor/experimental/test_local_map.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+        col_wise = [Shard(1)]  # col-wise sharding placements on 1-d mesh
+        W_dt = distribute_tensor(
+            W, device_mesh, col_wise
+        )  # col-wisely sharded W tensor
+        X_dt = distribute_tensor(
+            X, device_mesh, row_wise
+        )  # row-wisely sharded X tensor
+        # get the function wrapped with DTensor/Tensor convertion
+        # mm_allreduce_forward is a function that applies to Tensors with manual collective
+        # d_mm_allreduce_forward is the function that does the same but applies to DTensors
```
Comment (wanchaol) [inline test/distributed/_tensor/experimental/test_local_map.py:65]: ditto: now these funcs should be called as "local_mm_allreduce_forward"

File: torch/distributed/_tensor/experimental/local_map.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+from typing import Callable, Optional, Sequence, Tuple, Union
+
+import torch
+from torch.distributed._tensor import DeviceMesh, DTensor
+from torch.distributed._tensor.placement_types import Placement
+from torch.utils._pytree import tree_flatten, tree_unflatten
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/local_map.py:7]:
```
try:
    from torch.utils import _cxx_pytree as pytree
except ImportError:
    from torch.utils import _pytree as pytree  # type: ignore[no-redef]
```
This would make the pytree more efficient

File: torch/distributed/_tensor/experimental/local_map.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    Returns:
+        A `Callable` that applies `func` to each local shard of the input :class:`DTensor`
+        and returns a :class:`DTensor` constructed from the return value of `func`.
+
+    Raises:
+        AssertionError: If the input :class:`DTensor`s are not placed on the same device
+        mesh, or if they are placed on a different device mesh than the `device_mesh`
+        argument passed in.
+
+        AssertionError: If the placements in `out_placements` for non-:class:`DTensor`
```
Comment (wanchaol) [inline torch/distributed/_tensor/experimental/local_map.py:64]: I don't quite get this, could you rephrase this a bit to make it easier to understand?

**Review by wanchaol — APPROVED** [review]
lgtm! have a few comments inlined

File: test/distributed/_tensor/experimental/test_local_map.py (hunk 6 above)
Comment (XilunWu) [reply test/distributed/_tensor/experimental/test_local_map.py:20]: the motivation of limiting it to CPU test is that I think the correctness `local_map` should be consistent on CPU and GPU therefore we only need to test it on CPU.

File: torch/distributed/_tensor/experimental/local_map.py (hunk 7 above)
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/local_map.py:83]: my understanding of the naming is:
`local_map` turns a "local" function into a function on DTensor. The new function should be a "d" function. WDYT?

File: torch/distributed/_tensor/experimental/local_map.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+from torch.distributed._tensor.placement_types import Placement
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+
+PlacementType = Optional[Sequence[Placement]]
+InputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+OutputPlacements = Union[PlacementType, Tuple[PlacementType, ...]]
+
+
+def local_map(
```
Comment (yoyoyocmu) [inline torch/distributed/_tensor/experimental/local_map.py:15]: I think we can have a better function name once the before the release to prod.

File: torch/distributed/_tensor/experimental/local_map.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    func: Callable,
+    out_placements: OutputPlacements,
+    in_placements: Optional[InputPlacements] = None,
+    device_mesh: Optional[DeviceMesh] = None,
+    *,
+    redistribute_inputs: bool = False,
+):
```
Comment (yoyoyocmu) [inline torch/distributed/_tensor/experimental/local_map.py:22]: I have a use case that in the forward pass, the output of sub-module is of placement: `Replicate()`, but in the backward pass we want the grad's placement to be `_Partial()` to ensure numerical correctness. Wonder in the current API design, are we able to handle it?

File: torch/distributed/_tensor/experimental/local_map.py (hunk 7 above)
Comment (wanchaol) [reply torch/distributed/_tensor/experimental/local_map.py:83]: hmmm isn't it the reverse? 

`local_map` turns a function to a function that escapes DTensor and directly operate on local_tensor?

File: torch/distributed/_tensor/experimental/local_map.py (hunk 12 above)
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/local_map.py:15]: any recommendation on the naming? ;-)

File: torch/distributed/_tensor/experimental/local_map.py (hunk 7 above)
Comment (XilunWu) [reply torch/distributed/_tensor/experimental/local_map.py:83]: after offline sync, we decide the naming:
1. `local_map` means this function takes in a `func` that computes on `torch.Tensor` and maps this `func` to its `DTensor` inputs' `_local_tensor`.
2. The return value of `local_map` is a function which 1) performs redistribution on `DTensor` inputs when necessary according to `input_placements`; 2) calls `func` on each `DTensor._local_tensor`; 3) wrap the return value back to `DTensor` by `DTensor.from_local(local_tensor, device_mesh, out_placement)`. This returned function should have prefix `local_` to emphasize that it operates on `DTensor`'s local tensor.
